In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import cv2
from imblearn.over_sampling import SMOTE
from sklearn.model_selection import train_test_split

In [ ]:
import os
import cv2
import glob
import numpy as np
def getYourFruits(fruits, data_type, print_n=False, k_fold=False):
    images = []
    labels = []
    val = ['training', 'testing']
    if not k_fold:
        path = "/content/gdrive/MyDrive/AIL_project/traffic/Final Dataset/" + data_type + "/"
        for i,f in enumerate(fruits):
            p = path + f
            j=0
            for image_path in glob.glob(os.path.join(p, "*.jpg")):
                image = cv2.imread(image_path, cv2.IMREAD_COLOR)
                image = cv2.resize(image, (100, 100))
                image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)
                images.append(image)
                labels.append(i)
                j+=1
            if(print_n):
                print("There are " , j , " " , data_type.upper(), " images of " , fruits[i].upper())
        images = np.array(images)
        labels = np.array(labels)
        return images, labels
    else:
        for v in val:
            path = "/content/gdrive/MyDrive/AIL_project/traffic/Final Dataset/" + v + "/"
            for i,f in enumerate(fruits):
                p = path + f
                j=0
                for image_path in glob.glob(os.path.join(p, "*.jpg")):
                    image = cv2.imread(image_path, cv2.IMREAD_COLOR)
                    image = cv2.resize(image, (100, 100))
                    image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)
                    images.append(image)
                    labels.append(i)
                    j+=1
        images = np.array(images)
        labels = np.array(labels)
        return images, labels

def getAllFruits():
    fruits = []
    for fruit_path in glob.glob("/content/gdrive/MyDrive/AIL_project/traffic/Final Dataset/training/*"):
        fruit = fruit_path.split("/")[-1]
        fruits.append(fruit)
    return fruits
from sklearn.preprocessing import StandardScaler
fruits = ['Empty', 'Low' , 'Medium', 'High', 'Traffic Jam']


#Get Images and Labels
X, y =  getYourFruits(fruits, 'training')
X_test, y_test = getYourFruits(fruits, 'testing')

#Scale Data Images
scaler = StandardScaler()
X_train = scaler.fit_transform([i.flatten() for i in X])
X_test = scaler.fit_transform([i.flatten() for i in X_test])

In [ ]:
y_test

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4,
       4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
       4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,

**SMOTE**

**SMOTE** là một thuật toán  giúp giải quyết vấn đề dữ liệu mất cân bằng ( trong bài toán phân loại)

In [ ]:
from collections import Counter
counter = Counter(y)
print('Before', counter)
smt = SMOTE()
X_train_sm , y_train_sm = smt.fit_resample(X_train, y)
counter = Counter(y_train_sm)
print('After: ', counter)



Before Counter({0: 1186, 1: 946, 2: 688, 3: 378, 4: 181})
After:  Counter({0: 1186, 1: 1186, 2: 1186, 3: 1186, 4: 1186})


**SVM**

In [ ]:

from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

clf = SVC(kernel='linear', gamma = 'auto')
clf.fit(X_train_sm , y_train_sm)

SVC(gamma='auto', kernel='linear')

In [ ]:
y_pred = clf.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)

In [ ]:
accuracy

0.6625

--> Với cùng kernel và gamma giống với model SVM sử dụng trong file train model không sử dụng SMOTE. Thì sau khi sử dụng SMOTE, đã giúp cho hiệu suất của model tăng lên. Điều này chứng minh rằng, phương pháp SMOTE giúp giải quyết vấn đề dữ liệu mất cân bằng, từ đó làm tăng hiệu suất của model.

**KNN**

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
model_knn = KNeighborsClassifier(n_neighbors=4)
model_knn.fit(X_train_sm , y_train_sm)
y_pred_knn = model_knn.predict(X_test)

In [ ]:
accuracy = accuracy_score(y_test, y_pred_knn)
accuracy


0.6125

--> Với cùng số n_neighbors giống với model KNN sử dụng trong file train model không sử dụng SMOTE. Thì sau khi sử dụng SMOTE, đã giúp cho hiệu suất của model tăng lên. Cụ thể là : Khi chưa sử dụng SMOTE thì accuracy là : 0.42, còn sau khi sử dụng SMOTE accuracy của mô hình đã tăng lên 0.6125

**Conclusion**:
- Bằng cách sử dụng thuật toán SMOTE, đã giúp tăng hiệu suất của mô hình. Vì Smote giúp cân bằng tập dữ liệu.

-------
# **END**